presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java

```diff
@@ -319,18 +338,22 @@ private static MethodDefinition generateProcessMethod(
         return method;
     }
 
-    private static BytecodeBlock generateCheckYieldBlock(Variable thisVariable, Variable index, FieldDefinition yieldSignal, FieldDefinition nextIndexOrPosition)
+    private BytecodeBlock generateCheckYieldBlock(Variable thisVariable, Variable start, Variable index, FieldDefinition yieldSignal, FieldDefinition nextIndexOrPosition)
     {
         return new BytecodeBlock()
-                .comment("if (yieldSignal.isSet())")
+                .comment("if (System.nanoTime() - start > yieldCheckPeriodNanos)")
                 .append(new IfStatement()
-                        .condition(thisVariable.getField(yieldSignal).invoke("isSet", boolean.class))
+                        .condition(greaterThan(subtract(invokeStatic(System.class, "nanoTime", long.class), start), constantLong(yieldCheckPeriodNanos)))
```

Comment:
you can just and these together with `io.airlift.bytecode.expression.BytecodeExpressions#and`